src/main/java/de/tum/in/www1/artemis/repository/TextBlockRepository.java

```diff
@@ -30,4 +32,45 @@
     Set<TextBlock> findAllBySubmissionIdIn(Set<Long> submissionIdList);
 
     void deleteAllBySubmission_Id(Long submissionId);
+
+    /**
+     * Interface to define return type for countOtherBlocksInClusterForSubmission
+     */
+    interface TextBlockCount {
+
+        String getBlockId();
+
+        Long getNumberOfOtherBlocks();
+    }
+
+    /**
+     * For the given Submission `id` returns a list of raw object array representing two columns.
+     * First index/column corresponds to the TextBlock `id` while the second one corresponds to
+     * the number of other blocks in the same cluster as given block with id = `id`.
+     * For all TextBlock's of the Submission with the given `id`
+     * finds their respective cluster and retrieves the number of other blocks in the same cluster
+     * @param id the id of the Submission
+     * @return the number of other TextBlock's in the same cluster as the block with given `id`
+     */
+    @Query("""
+            SELECT tb.id as blockId, COUNT(DISTINCT tball.id) as numberOfOtherBlocks
+            FROM TextSubmission s
+            LEFT JOIN TextBlock tb ON s.id = tb.submission.id
+            LEFT JOIN TextCluster tc ON tb.cluster.id = tc.id
+            LEFT JOIN TextBlock tball ON tc.id = tball.cluster.id AND tball.id <> tb.id
+            WHERE s.id = :#{#id}
+            GROUP BY tb.id
+            """)
+    List<TextBlockCount> countOtherBlocksInClusterForSubmission(@Param("id") Long id);
+
+    /**
+     * This function calls query `countOtherBlocksInClusterForSubmission` and converts the result into a Map
+     * so that it's values will be easily accessed through key value pairs
+     * @param id the `id` of the Submission
+     * @return a Map data type representing key value pairs where the key is the TextBlock id
+     * and the value is the number of other blocks in the same cluster for that TextBlock.
+     */
+    default Map<String, Integer> countOtherBlocksInClusterForSubmissionByTextBlockId(Long id) {
```

Comment:
the method name sounds wrong: `ByTextBlockId` sounds like you would use a `textBlockId` as parameter, but it seems you use a `submissionId`
Please try to improve the method name

`id`-->`submissionId`